platform/security/session-management-endpoint/src/main/java/org/codice/ddf/security/session/management/impl/SessionManagementEndpointImpl.java

```diff
@@ -0,0 +1,151 @@
+      if (securityToken instanceof SecurityTokenHolder) {
+        timeLeft = getTimeLeft((SecurityTokenHolder) securityToken);
+      }
+    }
+    return new ByteArrayInputStream(Long.toString(timeLeft).getBytes(StandardCharsets.UTF_8));
+  }
+
+  public InputStream getRenewal(HttpServletRequest request) {
+    HttpSession session = request.getSession(false);
+    Response[] response = new Response[1];
+    InputStream inputStream = null;
+    if (session != null) {
+      Object securityToken = session.getAttribute(SecurityConstants.SAML_ASSERTION);
+      if (securityToken instanceof SecurityTokenHolder) {
+        SecurityTokenHolder tokenHolder = (SecurityTokenHolder) securityToken;
+        Map<String, SecurityToken> realmTokenMap = tokenHolder.getRealmTokenMap();
+        realmTokenMap
+            .keySet()
+            .forEach(
+                s -> {
+                  try {
+                    doRenew(s, realmTokenMap.get(s), tokenHolder);
+                  } catch (SecurityServiceException e) {
+                    response[0] = Response.serverError().build();
+                    LOGGER.error("Failed to renew", e);
+                  }
+                });
+        if (response[0] != null) {
+          return null;
+        }
+
+        inputStream =
+            new ByteArrayInputStream(
+                Long.toString(getTimeLeft(tokenHolder)).getBytes(StandardCharsets.UTF_8));
+      }
+    }
+    return inputStream;
+  }
+
+  public URI getInvalidate(HttpServletRequest request) {
+    StringBuffer requestURL = request.getRequestURL();
+    String requestQueryString = request.getQueryString();
+    return URI.create(
+        requestURL
+            .substring(0, requestURL.indexOf(rootContext))
+            .concat("/logout?noPrompt=true")
+            .concat(requestQueryString != null ? "&" + requestQueryString : ""));
+  }
+
+  private long getTimeLeft(SecurityTokenHolder securityToken) {
+    List<SecurityAssertionImpl> values =
+        securityToken
+            .getRealmTokenMap()
+            .values()
+            .stream()
+            .map(SecurityAssertionImpl::new)
+            .collect(Collectors.toList());
+    values.sort(
+        (o1, o2) -> {
+          long l = o1.getNotOnOrAfter().getTime() - o2.getNotOnOrAfter().getTime();
+          if (l > 0) {
+            return 1;
+          } else if (l < 0) {
+            return -1;
+          } else {
+            return 0;
+          }
+        });
+    return values.isEmpty()
+        ? 0
+        : Math.max(values.get(0).getNotOnOrAfter().getTime() - clock.millis(), 0);
+  }
+
+  private void doRenew(String realm, SecurityToken securityToken, SecurityTokenHolder tokenHolder)
+      throws SecurityServiceException {
+    SAMLAuthenticationToken samlToken =
+        new SAMLAuthenticationToken(securityToken.getPrincipal(), securityToken, realm);
+    Subject subject = securityManager.getSubject(samlToken);
```

Comment:
❗️ The `session-management-endpoint` bundle will need the same permissions as the `security-servlet-session-expiry` bundle. `catalog-ui-search` might need permissions as well if it doesn't already have them

Reply:
I wasn't seeing any access denied problems but I'll go ahead and investigate that. Do you know why it needs access to the keystore?

Reply:
This method calls out to the STS to renew the SAML assertion, and the STS has to sign the updated assertion which requires access to the keystores. I think that might be why.

The renewal only happens when the SAML assertion is about to expire, not when the user clicks "Continue working". By default, SAML assertions are valid for 30 minutes, so that might be why you don't see it. Try decreasing the SAML Assertion Lifetime or stop the CSRF bundle and hit the endpoint directly.

Reply:
Doesn't this mean the STS should be updated to have a privileged block if it's the one that needs access?

Reply:
If we used a privileged block, than any bundle could call to the STS to renew an assertion. Maybe that's what we want, but we've been very conservative about the use of privileged blocks to avoid confused deputy situations. I agree it's weird to give access to the keystore when the bundle doesn't need it per se, but at least this way we know that the bundle is trusted and isn't abusing the permission before granting it.